Fix float conversion from strings using comma as decimal separator

eternoendless opened the pull request:
<!-- Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information: -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | After editing the price of a combination product, the other combinations lose the decimal part from their prices, but only if the current locale uses comma as decimal separator.<br>This is because when PHP type-casts to int/float, it parses the string until it finds an invalid character. For example, both `echo (float) '123$456'` and `echo (float) '123,456'` yield `123`.<br>This fix introduces a new class that converts any comma into a dot so that the amount can be safely converted to float.
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/BOOM-2126
| How to test?  | See the ticket description

eternoendless committed:
BO: Fix float conversion from strings using comma as decimal separator

eternoendless committed:
BO: Remove unnecessary line

drAlberT commented on src/PrestaShopBundle/Utils/ImmutableFloat.php:
```diff
@@ -0,0 +1,58 @@
+    public function __construct($value)
+    {
+        if (!is_float($value)) {
+            throw new \InvalidArgumentException('Invalid argument: "%s" is not a valid float', $value);
```
missing `sprintf()`

eternoendless replied:
True! my bad

drAlberT commented on src/PrestaShopBundle/Utils/ImmutableFloat.php:
```diff
@@ -0,0 +1,58 @@
+    {
+        $value = (string) $value;
+        if ($value === '') {
+            return new self(0.0);
```
As the constructor is public and the class is not `final` IMHO you should use `static` instead of `self`

eternoendless replied:
Good point

drAlberT commented on src/PrestaShopBundle/Utils/ImmutableFloat.php:
```diff
@@ -0,0 +1,58 @@
+            return new self(0.0);
+        } else {
+            // ensure decimals are dot-separated instead of comma-separated
+            $value = str_replace(',', '.', $value);
```
This is not the right way to handle it .. what happens if I pass `1.024,99` for example?

drAlberT replied:
BTW:
- you can avoid the `else` too .. making code more readable
- you could follow [Yoda conditions](https://en.wikipedia.org/wiki/Yoda_conditions) in comparison

drAlberT commented on src/PrestaShopBundle/Utils/ImmutableFloat.php:
```diff
@@ -0,0 +1,58 @@
+/**
+ * Immutable value-object that contains a float value.
+ */
+class ImmutableFloat
```
IMVHO the `Immutable` part is tautological .. A float class is a VO, so it is obviously immutable. Avoiding `Immutable` makes it more compact, readable and easy to use .. If a MutableFloat would exists then I'll agree .. but having only one I think this is to be avoided

eternoendless replied:
Originally I started out with `Float`. Then two things crossed my mind:

1. `DateTime` and `DateTimeImmutable`. The immutable one has the word "immutable".
2. Scalar types on PHP >= 7. Consider the following function signature:
    
    ```php
    function doThing(float $bar): Float
    ```
    Namespaces aside, wouldn't it be confusing to have `Float` and `float`?

In addition, from the reader standpoint, if I saw something like this:

```php
$number = Float::fromString($string);
```

The first thing I would think would be "why did this guy implement his own Float type?", and then I would have to reverse-engineer the developer's reasoning to find out why. In my opinion, having the `Immutable` word makes it easier to understand at a glance – it's not about having a java-like Float type that tries to replace the native scalar type, it's about being sure that it cannot be modified and that it will always be valid.

If it was a "real" VO in VO namespace (for example if it was `VO\Amount`), I agree that having the word "Immutable" would be tautological. But since it's so generic, I believe that immutability cannot be taken for granted – you _need_ to read the phpdoc to know.

drAlberT replied:
a "real VO" has to be in a VO namespace? O_o Evans knows this?

The conflict between float and Float is not IMHO a reason to add Immutable .. (moreover, core classes such as \DateTime have Immutable as suffix .. )

Moreover you are putting something resembling a Type, a VO, a Class into the "Utils" namespace .. I expect a util class will perform actions (and your class target seems to do, as it want to perform string parsing) .. in conclusion IMVHO there are naming issues in your PR, and names are important to make the code clear, understandable and less error prone.

It's my own opinion of course

eternoendless replied:
> a "real VO" has to be in a VO namespace? O_o Evans knows this?

LOL I meant that as a "proper" VO having domain sense (it's a figure of speech). I don't think a Float VO makes sense from a domain standpoint because it's as generic as it gets.

> The conflict between float and Float is not IMHO a reason to add Immutable

Why not?

> moreover, core classes such as \DateTime have Immutable as suffix

So if it was FloatImmutable it would be OK?

> you are putting something resembling a Type, a VO, a Class into the "Utils" namespace

I agree that Utils may not be the best place to put it. What do you suggest?

drAlberT replied:
> > moreover, core classes such as \DateTime have Immutable as suffix

> So if it was FloatImmutable it would be OK?

No/Yes .. if your last decision is to reatain the Immutable world in the name I think it is just more consistent .. but my approach is the one that follows

> > you are putting something resembling a Type, a VO, a Class into the "Utils" namespace

> I agree that Utils may not be the best place to put it. What do you suggest?

Neglet the \Numberformatter part, something as exposed here (make the class name an "ActionMaker", and utility .. not a "type"): https://github.com/PrestaShop/PrestaShop/pull/7934#pullrequestreview-42821140

drAlberT reviewed (changes requested):
I started review as I found a missing `sprintf` .. then started some "personal" considerations too ;)

xBorderie commented:
Thanks for taking the time to make a code review, @drAlberT !
@eternoendless, what do you think of it?

eternoendless committed:
BO: Improve ImmutableFloat

- add missing sprintf
- use static instead of self
- use yoda conditions
- remove unneeded else

eternoendless committed:
BO: Improve ImmutableFloat

- support arbitrary string format
- add unit test

eternoendless commented:
Thanks for your review @drAlberT !

I have taken your comments into account and added support for arbitrary number formatting (see the unit test)

aleeks commented on src/PrestaShopBundle/Utils/ImmutableFloat.php:
```diff
@@ -0,0 +1,81 @@
+    public static function fromString($value)
+    {
+        $value = trim((string) $value);
+        if ('' === $value) {
```
what happens if $value = false or null ?

aleeks reviewed (changes requested):
Hello @eternoendless 
Cool for your tests! Can you rebase renaming your second commit please? missing `BO` :p

drAlberT reviewed (changes requested):
This is a wheel reinvenction IMHO .. please take into account the possibility to use \NumberFormatter

You should provide a "utility class" wrapping \Numberformatter in such a way that ease the use to the final user. Not a new "Float type" being able to do formatting/parsing, but a FloatParser tool 

something similar to 
echo FloatParser::fromString($theNumberHere);

My review is basically on the semantic of the class and the implementation of the formatting itself.

eternoendless committed:
BO: Improve ImmutableFloat

- make ::fromString more strict regarding what can be considered a number
- test more edge cases

eternoendless commented:
> This is a wheel reinvenction IMHO .. please take into account the possibility to use \NumberFormatter

We cannot use NumberFormatter because it's locale aware, so it only works if you know the locale of the number you're trying to transform beforehand. Take this example:

```php
// say this value comes form an <input type="number">
$value = '1234.567';
// and my current locale is 'fr_FR'
$currentLocale = 'fr_FR';

$n = new \NumberFormatter($currentLocale, \NumberFormatter::DECIMAL);
var_dump($n->parse($value));

// output: bool(false)
```
This doesn't work even if '1234.567' is a perfectly valid float, because the French format uses `,` as decimal separator. This means that even though my input remains the same, my transformation will sometimes fail if the input doesn't match my current locale.

The purpose of this class is precisely to cover that use case:
- It takes an arbitrarily-formatted number
- Makes sure it can be converted to float in the least ambiguous way possible
- Once an instance of it exists, we know for sure that it is valid

Of course, the better solution would be to make sure that only float-parsable strings are passed around, and reject everything else... that's the objective, but we aren't there yet.

eternoendless committed:
BO: Change ImmutableFloat into FloatParser

eternoendless committed:
BO: Harmonize string to float conversion on the product page

+ change price input on combinations to text in order to prevent problems when typing values with comma

drAlberT commented on src/PrestaShopBundle/Utils/FloatParser.php:
```diff
@@ -81,16 +61,6 @@ public static function fromString($value)
-    public function getValue()
-    {
-        return $this->value;
+        return (float) $value;
```
missing `$thi`s ?

drAlberT reviewed (approved):
Good work :)

I approve, but pay attention to the getValue() .. it lacks the `$this->`

drAlberT commented:
Sorry for boring .. but I think the  `fromString()` should be static .. no need to instantiate an object AFAICS

eternoendless committed:
BO: Cosmetic code fixes

eternoendless committed:
BO: Better handling of floats on combinations

The following fields are now parsed using arbitrary comma separator:
- "impact on weight"
- "impact on price per unit"

eternoendless commented:
> I think the fromString() should be static

As a general rule I think static classes are bad practise, because they prevent dependency injection.

drAlberT commented:
As a general rule a static *method* has no dependency at all  ;) ..  in this case it depends only on its argument .. a string

eternoendless commented:
I meant that if `FloatParser` is static (and it is considered static if its methods are static), it cannot be injected as a dependency.

So I think it's generally not a good idea to introduce static methods -- with the exception of factory methods, of course.

drAlberT commented:
O_o

This class has only a static method so it is "static" .. ok .. and so? this means it is only a method grouping class .. it can't be injected because there will make no sense to do at all .. it is only a tool to act on something else and returning something.. if you want / need to inject something in something-else then you have to inject the result of the static method.

IMHO your argument is over-killing .. the only reason I see for wanting to pass the object is mocking in tests, right now .. but, again, instead of mocking an object you could simply pass a fake string miming the result of the static method.

My 2 cents

eternoendless committed:
BO: Add JsDoc on priceCalculation

eternoendless committed:
BO: Chain event listeners on product combinations

+ remove dead code

eternoendless committed:
BO: Code cleanup

eternoendless committed:
BO: Fix impact on price not updated on the list

When the impact on price (with tax) was modified on the combination details, the impact on price (without tax) wasn't being updated on the combinations list

vincentbz commented:
![obamabeer](https://user-images.githubusercontent.com/7567060/27433956-1c336dc4-5757-11e7-9cae-8cabbbf8abfd.jpg)

eternoendless commented:
Had to rebase again due merge conflicts

maximebiloe commented:
Thank you @eternoendless and all the reviewers and testers

folkertsmit commented:
Hi i try to fix this problem, i have changed the 10 files but i get blank screen backend:
Parse error: syntax error, unexpected end of file in domain/classes/controller/AdminController.php on line 1284

when i restore the AdminController.php my backend works again. 

only change is line 920
+     * @return ObjectModel|bool

what went wrong?

seb-celinedesign commented:
Still the same issue in 1.7.2.3 version with BO language in french.
